components/org.wso2.carbon.identity.oauth2.grant.mfa.endpoints/src/main/java/org/wso2/carbon/identity/oauth2/grant/mfa/endpoints/impl/AuthenticationApiServiceImpl.java

```diff
@@ -0,0 +1,159 @@
+            return EndpointUtils.handleServerErrorResponse(authenticator, e, LOG);
+        } catch (Throwable e) {
+            return EndpointUtils.handleUnexpectedServerError(authenticator, e, LOG);
+        }
+    }
+
+    /**
+     * @param userAuthenticationRequest 
+     * @param userTenantDomain
+     * @return
+     */
+    @Override
+    public Response authenticatePost(UserAuthenticationRequest userAuthenticationRequest, String userTenantDomain) {
+
+        String userIdentifier = StringUtils.trim(userAuthenticationRequest.getUserIdentifier());
+        String password = StringUtils.trim(userAuthenticationRequest.getPassword());
+        String clientId = StringUtils.trim(userAuthenticationRequest.getClientId());
+        String flowId = StringUtils.trim(userAuthenticationRequest.getFlowId());
+        String authenticator = StringUtils.trim(userAuthenticationRequest.getAuthenticator());
+        UserAuthenticationResponseDTO responseDTO = null;
+
+        try {
+
+            if (StringUtils.isNotEmpty(StringUtils.trim(userAuthenticationRequest.getClientId())) &&
+                    StringUtils.isEmpty(StringUtils.trim(userAuthenticationRequest.getFlowId())) &&
+                    StringUtils.isNotEmpty(StringUtils.trim(userAuthenticationRequest.getUserIdentifier()))) {
+
+                responseDTO = EndpointUtils.getMFAAuthService().authenticateWithClientId(clientId, authenticator,
+                        password, userIdentifier, StringUtils.trim(userTenantDomain));
+
+            } else if (StringUtils.isNotEmpty(StringUtils.trim(userAuthenticationRequest.getFlowId())) &&
+                    StringUtils.isEmpty(StringUtils.trim(userAuthenticationRequest.getClientId()))) {
+
+                responseDTO = EndpointUtils.getMFAAuthService().authenticateWithFlowId
+                        (flowId, authenticator, password);
+            }
+
+            AuthenticationValidationFailureReasonDTO failureReasonDTO = responseDTO.getFailureReason();
+            AuthenticationValidationFailureReason failureReason = null;
+            if (failureReasonDTO != null) {
+                failureReason = new AuthenticationValidationFailureReason()
+                        .code(failureReasonDTO.getCode())
+                        .message(failureReasonDTO.getMessage())
+                        .description(failureReasonDTO.getDescription());
+            }
+
+            UserAuthenticationResponse response = new UserAuthenticationResponse()
+                    .isValidPassword(responseDTO.isValidPassword())
+                    .flowId(responseDTO.getFlowId())
+                    .authenticatedSteps(responseDTO.getAuthenticatedSteps())
+                    .isAuthFlowCompleted(responseDTO.isAuthFlowCompleted())
+                    .setAutheticatorStepDetails(responseDTO.getAuthenticationStepDetails())
+                    .nextStep(responseDTO.getNextStep())
+                    .failureReason(failureReason);
```

Comment:
In a successful case we do not have to set a failure reason. I remember @janakamarasena suggested that not to send null values in api response.